tools/cndump/cndb_record.c

```diff
@@ -0,0 +1,268 @@
+            indent, rec_type_name, r->name, r->cnid, r->cp.pfx_len,
+            r->cp.kvs_ext01 ? 'y' : 'n', reclen);
+        break;
+    }
+
+    case CNDB_TYPE_KVS_DEL: {
+        const struct cndb_rec_kvs_del *r = &rec->rec.kvs_del;
+        printf("%*s cnid %lu reclen %zu\n", indent, rec_type_name, r->cnid, reclen);
+        break;
+    }
+
+    case CNDB_TYPE_TXSTART: {
+        const struct cndb_rec_txstart *r = &rec->rec.txstart;
+        printf("%*s txid %lu seqno %lu ingestid %lu txhorizon %lu add %u del %u reclen %zu\n",
+            indent, rec_type_name, r->txid, r->seqno, r->ingestid,
+            r->txhorizon, r->add_cnt, r->del_cnt, reclen);
+        break;
+    }
+
+    case CNDB_TYPE_KVSET_ADD: {
+
+        const struct cndb_rec_kvset_add *r = &rec->rec.kvset_add;
+
+        printf("%*s txid %lu cnid %lu kvsetid %lu nodeid %lu dgen_hi %lu dgen_lo %lu "
+            "vused %lu compc %u reclen %zu",
+            indent, rec_type_name, r->txid, r->cnid, r->kvsetid, r->nodeid,
+            r->km.km_dgen_hi, r->km.km_dgen_lo, r->km.km_vused, r->km.km_compc, reclen);
+
+        print_long_line_break(oneline, indent);
+        printf("hblock 0x%lx", r->hblkid);
+
+        print_long_line_break(oneline, indent);
+        printf("kblocks %u", r->kblkc);
+        for (int i = 0; i < r->kblkc; i++)
+            printf(" 0x%lx", r->kblkv[i]);
+
+        print_long_line_break(oneline, indent);
+        printf("vblocks %u", r->vblkc);
+        for (int i = 0; i < r->vblkc; i++)
+            printf(" 0x%lx", r->vblkv[i]);
+
+        printf("\n");
+        break;
+    }
+
+    case CNDB_TYPE_KVSET_DEL: {
+        const struct cndb_rec_kvset_del *r = &rec->rec.kvset_del;
+        printf("%*s txid %lu cnid %lu kvsetid %lu reclen %zu\n",
+            indent, rec_type_name, r->txid, r->cnid, r->kvsetid, reclen);
+        break;
+    }
+
+    case CNDB_TYPE_KVSET_MOVE: {
+        const struct cndb_rec_kvset_move *r = &rec->rec.kvset_move;
+        printf("%*s cnid %lu src_nodeid %lu tgt_nodeid %lu reclen %zu",
+            indent, rec_type_name, r->cnid, r->src_nodeid, r->tgt_nodeid, reclen);
+        print_long_line_break(oneline, indent);
+        printf("kvsets %u", r->kvset_idc);
+        for (uint32_t i = 0; i < r->kvset_idc; i++)
+            printf(" %lu", r->kvset_idv[i]);
+        printf("\n");
+        break;
+    }
+
+    case CNDB_TYPE_ACK: {
+        const struct cndb_rec_ack *r = &rec->rec.ack;
+        char modified_name[32];
+        snprintf(modified_name, sizeof(modified_name), "%s%s",
+            rec_type_name, r->ack_type == CNDB_ACK_TYPE_ADD ? "_add" : "_del");
+        printf("%*s txid %lu cnid %lu kvsetid %lu reclen %zu\n",
+            indent, modified_name, r->txid, r->cnid, r->kvsetid, reclen);
+        break;
+    }
+
+    case CNDB_TYPE_NAK: {
+        const struct cndb_rec_nak *r = &rec->rec.nak;
+        printf("%*s txid %lu reclen %zu\n", indent, rec_type_name, r->txid, reclen);
+        break;
+    }
+    }
+}
+
+
```

Comment:
Not sure why a few extra newlines are so offensive.

Reply:
These empty newlines at the end of the file don't serve any purpose. I don't understand why they would be kept. Doesn't even help with "readability"

Reply:
Didn't people use to put CTRL-L in source file so that the paginator would start afresh when concatenating a list of files?  I vaguely recall something like that from way back...  Maybe it was for the printer, dunno...